src/Functions/array/arrayPrAUC.cpp

```diff
@@ -0,0 +1,241 @@
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            bool label = labels.getFloat64(current_offset + i) > 0;
+            sorted_labels[i].score = scores.getFloat64(current_offset + i);
+            sorted_labels[i].label = label;
+        }
+
+        /// Sorting scores in descending order to traverse the Precision Recall curve from left to right
+        std::sort(sorted_labels.begin(), sorted_labels.end(), [](const auto & lhs, const auto & rhs) { return lhs.score > rhs.score; });
+
+        size_t prev_tp = 0, curr_tp = 0;
+        size_t curr_p = 0;
+
+        Float64 prev_score = sorted_labels[0].score;
+        Float64 curr_precision = 1.0;
+
+        Float64 area = 0.0;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (sorted_labels[i].score != prev_score)
+            {
+                /* Precision = TP / (TP + FP)
+                 * Recall = TP / (TP + FN)
+                 *
+                 * Instead of calculating
+                 *   d_Area = Precision_n * (Recall_n - Recall_{n-1}),
+                 * we can just calculate
+                 *   d_Area = Precision_n * (TP_n - TP_{n-1})
+                 * and later divide it by (TP + FN).
+                 *
+                 * This can be done because (TP + FN) is constant and equal to total positive labels.
+                 */
+                curr_precision = static_cast<Float64>(curr_tp) / curr_p;
+                area += curr_precision * (curr_tp - prev_tp);
+                prev_tp = curr_tp;
+                prev_score = sorted_labels[i].score;
+            }
+
+            if (sorted_labels[i].label)
+                curr_tp += 1;
+            curr_p += 1;
+        }
+
+        curr_precision = curr_p > 0 ? static_cast<Float64>(curr_tp) / curr_p : 1.0;
+        area += curr_precision * (curr_tp - prev_tp);
+
+        /// If there were no positive labels, Recall did not change and the area is 0
+        if (curr_tp == 0)
+            return 0.0;
+        /// Finally, we divide by (TP + FN) to obtain the Recall
+        /// At this point we've traversed the whole curve and curr_tp = total positive labels (TP + FN)
+        return area / curr_tp;
+    }
+
+    static void vector(
+        const IColumn & scores,
+        const IColumn & labels,
+        const ColumnArray::Offsets & offsets,
+        PaddedPODArray<Float64> & result,
+        size_t input_rows_count)
+    {
+        result.resize(input_rows_count);
+
+        ColumnArray::Offset current_offset = 0;
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            auto next_offset = offsets[i];
+            result[i] = apply(scores, labels, current_offset, next_offset);
+            current_offset = next_offset;
+        }
+    }
+
+public:
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
```

Comment:
why 0?

Reply:
And if that's the case, what should the `arrayAUC`'s `getNumberOfArguments` return? It has two required args plus an optional arg.

Reply:
> I followed the current code on `arrayAUC`, which also returns 0. Should it be 2 for the `arrayPrAUC` instead?

yes, and regarding `arrayAUC`, I will look at it a bit later, after reviewing this function

Reply:
I think there is 0 for `arrayAUC` because the number of arguments is not determined. In our case it's always 2 so we can put 2. I've already changed it, don't worry.